WcaOnRails/app/models/add_new_result.rb

```diff
@@ -0,0 +1,397 @@
+
+    # set round_type_id
+    round = Round.find_by_id(round_id)
+    if round.cutoff.nil?
+      case round.number
+      when round.total_number_of_rounds
+        @round_type_id = "f"
+      when 1
+        @round_type_id = "1"
+      when 2
+        @round_type_id = "2"
+      when 3
+        @round_type_id = "3"
+      end
+    else
+      case round.number
+      when round.total_number_of_rounds
+        @round_type_id = "c"
+      when 1
+        @round_type_id = "d"
+      when 2
+        @round_type_id = "e"
+      when 3
+        @round_type_id = "g"
+      end
+    end
+    # set format_id
+    @format_id = round.format_id
+  end
+  
+  validate :require_competitor_to_have_not_competed_in_round
+  def require_competitor_to_have_not_competed_in_round
+    return unless errors.blank?
+    if is_new_competitor.to_i == 0 && Competition.find_by_id(competition_id).results.where(personId: competitor_id, eventId: event_id, roundTypeId: @round_type_id).any?
+      errors.add(:round_id, "Competitor currently has results for this round. To fix them use the Fix Results script")
+    end
+  end
+
+  validate :require_valid_value1
+  def require_valid_value1
+    if !value1.nil? && !@value1_solve_time.valid?
+      errors.add(:value1, "Not Valid")
+    end
+  end
+
+  validate :require_valid_value2
+  def require_valid_value2
+    if !value2.nil? && !@value2_solve_time.valid?
+      errors.add(:value2, "Not Valid")
+    end
+  end
+
+  validate :require_valid_value3
+  def require_valid_value3
+    if !value3.nil? && !@value3_solve_time.valid?
+      errors.add(:value3, "Not Valid")
+    end
+  end
+
+  validate :require_valid_value4
+  def require_valid_value4
+    if !value4.nil? && !@value4_solve_time.valid?
+      errors.add(:value4, "Not Valid")
+    end
+  end
+
+  validate :require_valid_value5
+  def require_valid_value5
+    if !value5.nil? && !@value5_solve_time.valid?
+      errors.add(:value5, "Not Valid")
+    end
+  end
+
+  def do_add_new_result
+    if !valid?
+      return { error: "invalid forum" }
+    end
+
+    ActiveRecord::Base.transaction do
+      new_person_res = create_new_person
+      if !new_person_res.nil? && new_person_res[:error]
+        return new_person_res
+      end
+
+      new_result_res = create_new_result
+      if !new_result_res.nil? && new_result_res[:error]
+        return new_result_res
+      end
+
+      fix_positions
+    end
+    
+    { wca_id: @use_wca_id }
+  end
+
+  private def generate_new_wca_id
+    # generate new wcaid
+    similarWcaIds = Person.where("wca_id LIKE ?", semi_id + '%')
+    (1..99).each do |i|
+      if !similarWcaIds.where(wca_id: semi_id + i.to_s.rjust(2, "0")).any?
+        @new_wca_id = semi_id + i.to_s.rjust(2, "0")
+        return true
+      end
+    end
+
+    # Semi Id doesn't work
+    false
+  end
+
+  # create new competitor if needed and set wca_id, name, and country_id needed for the result row
+  def create_new_person
+    if is_new_competitor.to_i == 1
```

Comment:
The only issue is if someone fills in the WCAID, then switches to create a new person and fills out the rest of the form, then whenever submitting all fields (competitor id and the name, dob, country, gender are filled in) and the only way to know if the user wanted to create a new competitor or use the id is the `is_new_competitor ` field.

Reply:
Oh I see, then I think it can be dealt with from the UI perspective:
  - start with a fully enabled empty form for user data
  - if the WCA ID field is filled with something looking like a WCA ID: fetch the person's data, fill the other fields with them, disable them (or display an error to the user if the WCA ID is wrong)
  - if the WCA ID is cleared, re-enable the other fields.
  - if the other fields start having data: clear and disable the WCA ID field.
  - if the other fields are cleared, re-enable the WCA ID one.

From the server perspective:
  - if receiving a WCA ID, ignore all other person's attribute and set the WCA ID's person as the one for the result.
  - if not, perform the creation/validation of a new person based on the person's attribute received.

I think from a user perspective it's clearer what data will be used to create/associate the person for the result, and from the server perspective it's straightforward to implement.